[novatel_oem7_msgs] what():  unordered_map::at

```
Starting >>> novatel_oem7_msgs
--- stderr: novatel_oem7_msgs
terminate called after throwing an instance of 'std::out_of_range'
  what():  unordered_map::at
gmake[2]: *** [CMakeFiles/novatel_oem7_msgs__rosidl_generator_type_description.dir/build.make:162: rosidl_generator_type_description/novatel_oem7_msgs/srv/Oem7AbasciiCmd.json] Aborted (core dumped)
gmake[1]: *** [CMakeFiles/Makefile2:215: CMakeFiles/novatel_oem7_msgs__rosidl_generator_type_description.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< novatel_oem7_msgs [1.48s, exited with code 2]
```

Steps to reproduce:

```Dockerfile
FROM ros:rolling
SHELL [ "/bin/bash", "-c" ]
RUN apt-get update -qq && apt-get install -qqy python3-colcon-core
RUN mkdir -p /ws/src
WORKDIR /ws
RUN git clone --branch=humble https://github.com/novatel/novatel_oem7_driver.git src/novatel_oem7_driver
RUN rosdep update && rosdep install -y --from-paths src/novatel_oem7_driver/src/novatel_oem7_msgs
# Optional: rosidlcpp, compiles only without it
RUN apt-get install -qqy ros-rolling-rosidlcpp
RUN source /opt/ros/*/setup.bash &&  colcon build --packages-select novatel_oem7_msgs
```

The content of the service does not really look suspicious:
[**Oem7AbasciiCmd.srv**](https://github.com/novatel/novatel_oem7_driver/blob/humble/src/novatel_oem7_msgs/srv/Oem7AbasciiCmd.srv) 
```idl
string cmd
---
string rsp
```

--- a/rosidlcpp_generator_c/src/rosidlcpp_generator_c/rosidlcpp_generator_c.cpp
+++ b/rosidlcpp_generator_c/src/rosidlcpp_generator_c/rosidlcpp_generator_c.cpp
@@ -699,9 +699,13 @@ void GeneratorC::run() {
     ros_json["raw_source"]["content"] = nlohmann::json::array();
     std::string line;
     while (std::getline(raw_source_stream, line)) {
+      // Handle windows-style line endings
+      if (!line.empty() && line.back() == '\r') {
+        line.pop_back();
+      }
       ros_json["raw_source"]["content"].push_back(line);
     }
-    ros_json["raw_source"]["encoding"] = std::filesystem::path(raw_source_file_path).extension().string().substr(1); // Remove the leading dot from the extension
+    ros_json["raw_source"]["encoding"] = std::filesystem::path(raw_source_file_path).extension().string().substr(1);  // Remove the leading dot from the extension
 
     ros_json["package_name"] = m_arguments.package_name;
 

--- a/rosidlcpp_generator_cpp/src/rosidlcpp_generator_cpp/rosidlcpp_generator_cpp.cpp
+++ b/rosidlcpp_generator_cpp/src/rosidlcpp_generator_cpp/rosidlcpp_generator_cpp.cpp
@@ -469,13 +469,15 @@ auto generate_default_string(const nlohmann::json &membset, const std::string &)
     if (member["default_value"].is_array()) {
       if (std::all_of(member["default_value"].begin(),
                       member["default_value"].end(),
-                      [first = member["default_value"]](const auto &value) {
+                      [first = member["default_value"].front()](const auto &value) {
                         return value == first;
                       })) {
-        result.push_back(
-            "this->" + member["name"].get<std::string>() + ".fill(" +
-            msg_type_only_to_cpp(member["type"]["name"]) + "{" +
-            member["default_value"][0].get<std::string>() + "});");
+        result.push_back(std::format("std::fill<typename {}::iterator, {}>(this->{}.begin(), this->{}.end(), {});",
+                                     msg_type_to_cpp(member["type"]),
+                                     msg_type_only_to_cpp(member["type"]),
+                                     member["name"].get<std::string>(),
+                                     member["name"].get<std::string>(),
+                                     member["default_value"][0].get<std::string>()));
       } else {
         for (size_t i = 0; i < member["default_value"].size(); ++i) {
           result.push_back("this->" + member["name"].get<std::string>() +

--- a/rosidlcpp_generator_py/resource/_msg.py.template
+++ b/rosidlcpp_generator_py/resource/_msg.py.template
@@ -163,7 +163,7 @@ string{% if existsIn(type_, "maximum_size") -%}
 {{ get_rosidl_parser_type(member.type) }}({% endif %}
 {{ get_rosidl_parser_type(type_) }}({% if is_primitive(type_) and not is_string(type_) %}
 '{{ type_.name }}'{% else if is_string(type_) and existsIn(type_, "maximum_size") %}
-{{ member.type.maximum_size -}}
+{{ type_.maximum_size -}}
 {% else if existsIn(type_, "namespaces") %}
 [{{ "'" + join(type_.namespaces, "', '") + "'" }}], '{{ type_.name }}'{% endif %}
 ){% if member.type.name == "array" %}

--- a/rosidlcpp_generator_type_description/src/rosidlcpp_generator_type_description/rosidlcpp_generator_type_description.cpp
+++ b/rosidlcpp_generator_type_description/src/rosidlcpp_generator_type_description/rosidlcpp_generator_type_description.cpp
@@ -156,6 +156,10 @@ auto field_type_type_name(const nlohmann::json &ftype) -> std::string {
     value_type_name = FIELD_VALUE_TYPE_MAP.at(value_type["name"].get<std::string>());
   }
 
+  if (value_type_name.empty()) {
+    throw std::runtime_error("Unknown field type: " + value_type.dump());
+  }
+
   return value_type_name + name_suffix;
 }
 

--- a/rosidlcpp_parser/include/rosidlcpp_parser/rosidlcpp_parser.hpp
+++ b/rosidlcpp_parser/include/rosidlcpp_parser/rosidlcpp_parser.hpp
@@ -39,7 +39,7 @@ auto parse_string(std::string_view& content_view) -> std::string;
 auto parse_string_part(std::string_view& content_view) -> std::string;
 auto parse_string_python(std::string_view& content_view) -> std::string;
 auto interpret_type(std::string_view content_view, TypedefMap typedefs = {}) -> nlohmann::json;
-auto parse_type(std::string_view& content_view) -> std::string_view;
+auto parse_type(std::string_view& content_view) -> std::string;
 auto parse_name(std::string_view& content_view) -> std::string_view;
 auto parse_value(std::string_view& content_view) -> nlohmann::json;
 auto parse_constant(std::string_view& content_view, TypedefMap typedefs = {}) -> nlohmann::json;

--- a/rosidlcpp_parser/src/rosidlcpp_parser/rosidlcpp_parser.cpp
+++ b/rosidlcpp_parser/src/rosidlcpp_parser/rosidlcpp_parser.cpp
@@ -87,6 +87,12 @@ auto consume_comment(std::string_view& content_view) -> void {
     } else {
       content_view.remove_prefix(end_of_line + 1);
     }
+  } else if (content_view.substr(0, 2) == "/*") {
+    auto end_of_comment = content_view.find("*/");
+    if (end_of_comment == std::string_view::npos) {
+      throw std::runtime_error("Unterminated comment in content view");
+    }
+    content_view.remove_prefix(end_of_comment + 2);
   }
 }
 
@@ -126,6 +132,11 @@ auto interpret_type(std::string_view type_string, TypedefMap typedefs) -> json {
     type_string.remove_prefix(std::string_view{"string<"}.size());
     type_string.remove_suffix(1);  // Remove last '>'
     result["maximum_size"] = std::stoi(std::string{remove_white_space(type_string)});
+  } else if (type_string.starts_with("wstring<")) {  // wstring<bounds>
+    result["name"] = "wstring";
+    type_string.remove_prefix(std::string_view{"wstring<"}.size());
+    type_string.remove_suffix(1);  // Remove last '>'
+    result["maximum_size"] = std::stoi(std::string{remove_white_space(type_string)});
   } else if (type_string.starts_with("sequence<")) {  // sequence<type> or sequence<type, bounds>
     result["name"] = "sequence";
 
@@ -162,14 +173,31 @@ auto interpret_type(std::string_view type_string, TypedefMap typedefs) -> json {
   return result;
 }
 
-auto parse_type(std::string_view& content_view) -> std::string_view {
+auto parse_type(std::string_view& content_view) -> std::string {
   auto end_of_type = content_view.find_first_not_of(VALID_TYPE_CHAR);
   if (end_of_type == std::string_view::npos) {
     throw std::runtime_error("Malformed type");
   }
 
-  if (content_view.at(end_of_type) == '<') {
-    end_of_type = content_view.find_first_of('>') + 1;
+  std::string type_name{content_view.substr(0, end_of_type)};
+  content_view.remove_prefix(end_of_type);
+  end_of_type = 0;
+
+  consume_white_space_and_comment(content_view);
+
+  if (content_view[0] == '<') {
+    auto tmp_end_of_type = content_view.find_first_of('>') + 1;
+
+    if (tmp_end_of_type == std::string_view::npos) {
+      throw std::runtime_error("Malformed type: missing '>'");
+    }
+
+    // Check if there is a '<' in the type. If there is, we need to find the next '>'
+    if (content_view.substr(end_of_type + 1, tmp_end_of_type - end_of_type).find_first_of('<') != std::string_view::npos) {
+      tmp_end_of_type = content_view.find_first_of('>', tmp_end_of_type) + 1;
+    }
+
+    end_of_type = tmp_end_of_type;
   }
 
   // Handle multi word types
@@ -190,10 +218,10 @@ auto parse_type(std::string_view& content_view) -> std::string_view {
     end_of_type = std::string_view("long double").size();
   }
 
-  auto type_string = content_view.substr(0, end_of_type);
+  type_name += content_view.substr(0, end_of_type);
   content_view.remove_prefix(end_of_type);
 
-  return type_string;
+  return type_name;
 }
 
 auto parse_value_list(std::string_view& content_view) -> json {
@@ -413,11 +441,13 @@ auto parse_member(std::string_view& content_view, TypedefMap typedefs) -> json {
 
   consume_white_space_and_comment(content_view);
 
-  // if (content_view[0] == '=') {
-  //     content_view.remove_prefix(1);
-  //     remove_white_space_and_comment(content_view);
-  //     result["default"] = parse_value(content_view);
-  // }
+  if (content_view[0] == '[') {  // Array definition
+    auto end_of_array_definition = content_view.find_first_of(']') + 1;
+    result["type"]["value_type"] = result["type"];
+    result["type"]["name"] = "array";
+    result["type"]["size"] = std::stoi(std::string{remove_white_space(content_view.substr(1, end_of_array_definition - 2))});
+    content_view.remove_prefix(end_of_array_definition);
+  }
 
   consume_white_space_and_comment(content_view);
 

--- a/rosidlcpp_parser/test/unit/rosidlcpp_parser_test.cpp
+++ b/rosidlcpp_parser/test/unit/rosidlcpp_parser_test.cpp
@@ -26,114 +26,225 @@
 using namespace rosidlcpp_parser;
 
 TEST(RosIdlParserTest, RemoveWhiteSpace) {
-    std::string_view test_string_1 = "   a";
-    std::string_view test_string_2 = "\ta";
-    std::string_view test_string_3 = "  \ta";
-    std::string_view test_string_4 = "b  \ta";
+  std::string_view test_string_1 = "   a";
+  std::string_view test_string_2 = "\ta";
+  std::string_view test_string_3 = "  \ta";
+  std::string_view test_string_4 = "b  \ta";
 
-    consume_white_space(test_string_1);
-    EXPECT_EQ(test_string_1, "a");
+  consume_white_space(test_string_1);
+  EXPECT_EQ(test_string_1, "a");
 
-    consume_white_space(test_string_2);
-    EXPECT_EQ(test_string_2, "a");
+  consume_white_space(test_string_2);
+  EXPECT_EQ(test_string_2, "a");
 
-    consume_white_space(test_string_3);
-    EXPECT_EQ(test_string_3, "a");
+  consume_white_space(test_string_3);
+  EXPECT_EQ(test_string_3, "a");
 
-    consume_white_space(test_string_4);
-    EXPECT_EQ(test_string_4, "b  \ta");
+  consume_white_space(test_string_4);
+  EXPECT_EQ(test_string_4, "b  \ta");
 }
 
 TEST(RosIdlParserTest, RemoveComment) {
-    std::string_view test_string_1 = "// This is a test comment";
-    std::string_view test_string_2 = "// This is a test comment\n// This is the next line";
-    std::string_view test_string_3 = "This is not a comment";
+  std::string_view test_string_1 = "// This is a test comment";
+  std::string_view test_string_2 = "// This is a test comment\n// This is the next line";
+  std::string_view test_string_3 = "This is not a comment";
 
-    consume_comment(test_string_1);
-    EXPECT_EQ(test_string_1, "");
+  consume_comment(test_string_1);
+  EXPECT_EQ(test_string_1, "");
 
-    consume_comment(test_string_2);
-    EXPECT_EQ(test_string_2, "// This is the next line");
+  consume_comment(test_string_2);
+  EXPECT_EQ(test_string_2, "// This is the next line");
 
-    consume_comment(test_string_3);
-    EXPECT_EQ(test_string_3, "This is not a comment");
+  consume_comment(test_string_3);
+  EXPECT_EQ(test_string_3, "This is not a comment");
 }
 
 TEST(RosIdlParserTest, RemoveWhiteSpaceAndComment) {
-    std::string_view test_string_1 = "  \t// This is a one line comment";
-    std::string_view test_string_2 = "  \t// This is another one line comment\nThis is not a comment";
-    std::string_view test_string_3 = "  // This is a multiline comment\n// This is the next line\nThis is not a comment";
+  std::string_view test_string_1 = "  \t// This is a one line comment";
+  std::string_view test_string_2 = "  \t// This is another one line comment\nThis is not a comment";
+  std::string_view test_string_3 = "  // This is a multiline comment\n// This is the next line\nThis is not a comment";
 
-    consume_white_space_and_comment(test_string_1);
-    EXPECT_EQ(test_string_1, "");
+  consume_white_space_and_comment(test_string_1);
+  EXPECT_EQ(test_string_1, "");
 
-    consume_white_space_and_comment(test_string_2);
-    EXPECT_EQ(test_string_2, "This is not a comment");
+  consume_white_space_and_comment(test_string_2);
+  EXPECT_EQ(test_string_2, "This is not a comment");
 
-    consume_white_space_and_comment(test_string_3);
-    EXPECT_EQ(test_string_3, "This is not a comment");
+  consume_white_space_and_comment(test_string_3);
+  EXPECT_EQ(test_string_3, "This is not a comment");
 }
 
 TEST(RosIdlParserTest, ParseName) {
-    std::string_view test_string_1 = "";
-    std::string_view test_string_2 = "abcABC123_";
-    std::string_view test_string_3 = "abcABC123_   ";
-    std::string_view test_string_4 = "abcAB-C123_";
+  std::string_view test_string_1 = "";
+  std::string_view test_string_2 = "abcABC123_";
+  std::string_view test_string_3 = "abcABC123_   ";
+  std::string_view test_string_4 = "abcAB-C123_";
 
-    EXPECT_EQ(parse_name(test_string_1), "");
-    EXPECT_EQ(parse_name(test_string_2), "abcABC123_");
-    EXPECT_EQ(parse_name(test_string_3), "abcABC123_");
-    EXPECT_EQ(parse_name(test_string_4), "abcAB");
-    EXPECT_EQ(test_string_4, "-C123_");
+  EXPECT_EQ(parse_name(test_string_1), "");
+  EXPECT_EQ(parse_name(test_string_2), "abcABC123_");
+  EXPECT_EQ(parse_name(test_string_3), "abcABC123_");
+  EXPECT_EQ(parse_name(test_string_4), "abcAB");
+  EXPECT_EQ(test_string_4, "-C123_");
 }
 
 TEST(RosIdlParserTest, ParseStructure) {
-    std::string_view test_string_1 = "struct EmptyStruct {};// After struct";
-    std::string_view test_string_2 = "struct EmptyStructWithSpace   \n\t  {\n\n\t   };// After struct";
+  std::string_view test_string_1 = "struct EmptyStruct {};// After struct";
+  std::string_view test_string_2 = "struct EmptyStructWithSpace   \n\t  {\n\n\t   };// After struct";
 
-    auto struct_result_1 = parse_structure(test_string_1);
-    ASSERT_TRUE(struct_result_1.contains("name"));
-    EXPECT_EQ(struct_result_1["name"], "EmptyStruct");
-    EXPECT_EQ(test_string_1, "");
+  auto struct_result_1 = parse_structure(test_string_1);
+  ASSERT_TRUE(struct_result_1.contains("name"));
+  EXPECT_EQ(struct_result_1["name"], "EmptyStruct");
+  EXPECT_EQ(test_string_1, "");
 
-    auto struct_result_2 = parse_structure(test_string_2);
-    ASSERT_TRUE(struct_result_2.contains("name"));
-    EXPECT_EQ(struct_result_2["name"], "EmptyStructWithSpace");
-    EXPECT_EQ(test_string_2, "");
+  auto struct_result_2 = parse_structure(test_string_2);
+  ASSERT_TRUE(struct_result_2.contains("name"));
+  EXPECT_EQ(struct_result_2["name"], "EmptyStructWithSpace");
+  EXPECT_EQ(test_string_2, "");
 }
 
 TEST(RosIdlParserTest, ParseValueString) {
-    std::string_view test_string_1 = "\"This is a test string\"Unparsed data";
-    std::string_view test_string_2 = "\"This is a test string\\\"with escaped elements\"Unparsed data";
-    std::string_view test_string_3 = "\"This is a test string\\\"with multiple\\\" escaped elements\"Unparsed data";
+  std::string_view test_string_1 = "\"This is a test string\"Unparsed data";
+  std::string_view test_string_2 = "\"This is a test string\\\"with escaped elements\"Unparsed data";
+  std::string_view test_string_3 = "\"This is a test string\\\"with multiple\\\" escaped elements\"Unparsed data";
 
-    EXPECT_EQ(parse_string(test_string_1), "This is a test string");
-    EXPECT_EQ(parse_string(test_string_2), "This is a test string\"with escaped elements");
-    EXPECT_EQ(parse_string(test_string_3), "This is a test string\"with multiple\" escaped elements");
+  EXPECT_EQ(parse_string(test_string_1), "This is a test string");
+  EXPECT_EQ(parse_string(test_string_2), "This is a test string\"with escaped elements");
+  EXPECT_EQ(parse_string(test_string_3), "This is a test string\"with multiple\" escaped elements");
 }
 
 TEST(RosIdlParserTest, ParseTypedef) {
-    std::string_view test_string_1 = "typedef uint8 other_name;";
-    std::string_view test_string_2 = "typedef uint8 other_name;\n";
+  std::string_view test_string_1 = "typedef uint8 other_name;";
+  std::string_view test_string_2 = "typedef uint8 other_name;\n";
 
-    std::pair<std::string, std::string> result_1 = {"other_name", "uint8"};
-    std::pair<std::string, std::string> result_2 = {"other_name", "uint8"};
+  std::pair<std::string, std::string> result_1 = {"other_name", "uint8"};
+  std::pair<std::string, std::string> result_2 = {"other_name", "uint8"};
 
-    EXPECT_EQ(parse_typedef(test_string_1), result_1);
-    EXPECT_EQ(parse_typedef(test_string_2), result_2);
+  EXPECT_EQ(parse_typedef(test_string_1), result_1);
+  EXPECT_EQ(parse_typedef(test_string_2), result_2);
+}
+
+TEST(RosIdlParserTest, ParseType) {
+  std::string_view test_string_1 = "uint8 member_name;";
+  std::string_view test_string_2 = "sequence<uint8> member_name;";
+  std::string_view test_string_3 = "sequence<uint8, 6> member_name;";
+  std::string_view test_string_4 = "string<6> member_name;";
+  std::string_view test_string_5 = "sequence<string> member_name;";
+  std::string_view test_string_6 = "sequence<string<6>> member_name;";
+  std::string_view test_string_7 = "sequence<string<6>, 10> member_name;";
+  std::string_view test_string_8 = "sequence<  string< 6 > > member_name;";
+  std::string_view test_string_9 = "string<4> member_name;";
+  std::string_view test_string_10 = "wstring<5> member_name;";
+  std::string_view test_string_11 = "sequence <  string< 6 > > member_name;";
+
+  std::string_view result_1 = "uint8";
+  std::string_view result_2 = "sequence<uint8>";
+  std::string_view result_3 = "sequence<uint8, 6>";
+  std::string_view result_4 = "string<6>";
+  std::string_view result_5 = "sequence<string>";
+  std::string_view result_6 = "sequence<string<6>>";
+  std::string_view result_7 = "sequence<string<6>, 10>";
+  std::string_view result_8 = "sequence<  string< 6 > >";
+  std::string_view result_9 = "string<4>";
+  std::string_view result_10 = "wstring<5>";
+  std::string_view result_11 = "sequence<  string< 6 > >";
+
+  EXPECT_EQ(parse_type(test_string_1), result_1);
+  EXPECT_EQ(parse_type(test_string_2), result_2);
+  EXPECT_EQ(parse_type(test_string_3), result_3);
+  EXPECT_EQ(parse_type(test_string_4), result_4);
+  EXPECT_EQ(parse_type(test_string_5), result_5);
+  EXPECT_EQ(parse_type(test_string_6), result_6);
+  EXPECT_EQ(parse_type(test_string_7), result_7);
+  EXPECT_EQ(parse_type(test_string_8), result_8);
+  EXPECT_EQ(parse_type(test_string_9), result_9);
+  EXPECT_EQ(parse_type(test_string_10), result_10);
+  EXPECT_EQ(parse_type(test_string_11), result_11);
+}
+
+TEST(RosIdlParserTest, InterpretType) {
+  std::string_view test_string_1 = "uint8";
+  std::string_view test_string_2 = "sequence<uint8>";
+  std::string_view test_string_3 = "sequence<uint8, 6>";
+  std::string_view test_string_4 = "string<6>";
+  std::string_view test_string_5 = "sequence<string>";
+  std::string_view test_string_6 = "sequence<string<6>>";
+  std::string_view test_string_7 = "sequence<string<6>, 10> member_name;";
+  std::string_view test_string_8 = "sequence<  string< 6 > > member_name;";
+  std::string_view test_string_9 = "string<4>";
+  std::string_view test_string_10 = "wstring<5>";
+
+  nlohmann::json result_1 = {{"name", "uint8"}};
+  nlohmann::json result_2 = {{"name", "sequence"}, {"value_type", {{"name", "uint8"}}}};
+  nlohmann::json result_3 = {{"name", "sequence"}, {"value_type", {{"name", "uint8"}}}, {"maximum_size", 6}};
+  nlohmann::json result_4 = {{"name", "string"}, {"maximum_size", 6}};
+  nlohmann::json result_5 = {{"name", "sequence"}, {"value_type", {{"name", "string"}}}};
+  nlohmann::json result_6 = {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}};
+  nlohmann::json result_7 = {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}, {"maximum_size", 10}};
+  nlohmann::json result_8 = {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}};
+  nlohmann::json result_9 = {{"name", "string"}, {"maximum_size", 4}};
+  nlohmann::json result_10 = {{"name", "wstring"}, {"maximum_size", 5}};
+  nlohmann::json result_11 = {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}};
+
+  EXPECT_EQ(interpret_type(test_string_1), result_1);
+  EXPECT_EQ(interpret_type(test_string_2), result_2);
+  EXPECT_EQ(interpret_type(test_string_3), result_3);
+  EXPECT_EQ(interpret_type(test_string_4), result_4);
+  EXPECT_EQ(interpret_type(test_string_5), result_5);
+  EXPECT_EQ(interpret_type(test_string_6), result_6);
+  EXPECT_EQ(interpret_type(test_string_7), result_7);
+  EXPECT_EQ(interpret_type(test_string_8), result_8);
+  EXPECT_EQ(interpret_type(test_string_9), result_9);
+  EXPECT_EQ(interpret_type(test_string_10), result_10);
 }
 
 TEST(RosIdlParserTest, ParseAttribute) {
-    std::string_view test_string_1 = "@key long key;";
-    std::string_view test_string_2 = "@default (value=1.23)\n";
-    std::string_view test_string_3 = "@verbatim (language=\"comment\", text=\"My comment\")\n";
+  std::string_view test_string_1 = "@key long key;";
+  std::string_view test_string_2 = "@default (value=1.23)\n";
+  std::string_view test_string_3 = "@verbatim (language=\"comment\", text=\"My comment\")\n";
 
-    nlohmann::json result_1 = {{"name", "key"}};
-    nlohmann::json result_2 = {{"name", "default"}, {"content", {{"value", 1.23}}}};
-    nlohmann::json result_3 = {{"name", "verbatim"}, {"content", {{"language", "comment"}, {"text", "My comment"}}}};
+  nlohmann::json result_1 = {{"name", "key"}};
+  nlohmann::json result_2 = {{"name", "default"}, {"content", {{"value", 1.23}}}};
+  nlohmann::json result_3 = {{"name", "verbatim"}, {"content", {{"language", "comment"}, {"text", "My comment"}}}};
 
-    EXPECT_EQ(parse_attribute(test_string_1), result_1);
-    EXPECT_EQ(parse_attribute(test_string_2), result_2);
-    EXPECT_EQ(parse_attribute(test_string_3), result_3);
+  EXPECT_EQ(parse_attribute(test_string_1), result_1);
+  EXPECT_EQ(parse_attribute(test_string_2), result_2);
+  EXPECT_EQ(parse_attribute(test_string_3), result_3);
 }
 
+TEST(RosIdlParserTest, ParseMember) {
+  std::string_view test_string_1 = "uint8 member_name;";
+  std::string_view test_string_2 = "sequence<uint8> member_name;";
+  std::string_view test_string_3 = "sequence<uint8, 6> member_name;";
+  std::string_view test_string_4 = "string<6> member_name;";
+  std::string_view test_string_5 = "sequence<string> member_name;";
+  std::string_view test_string_6 = "sequence<string<6>> member_name;";
+  std::string_view test_string_7 = "sequence<string<6>, 10> member_name;";
+  std::string_view test_string_8 = "sequence<  string< 6 > > member_name;";
+  std::string_view test_string_9 = "string<4> member_name;";
+  std::string_view test_string_10 = "wstring<5> member_name;";
+  std::string_view test_string_11 = "uint8 member_name[123];";
+
+  nlohmann::json result_1 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "uint8"}}}, {"name", "member_name"}};
+  nlohmann::json result_2 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "uint8"}}}}}, {"name", "member_name"}};
+  nlohmann::json result_3 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "uint8"}}}, {"maximum_size", 6}}}, {"name", "member_name"}};
+  nlohmann::json result_4 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "string"}, {"maximum_size", 6}}}, {"name", "member_name"}};
+  nlohmann::json result_5 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "string"}}}}}, {"name", "member_name"}};
+  nlohmann::json result_6 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}}}, {"name", "member_name"}};
+  nlohmann::json result_7 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}, {"maximum_size", 10}}}, {"name", "member_name"}};
+  nlohmann::json result_8 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "sequence"}, {"value_type", {{"name", "string"}, {"maximum_size", 6}}}}}, {"name", "member_name"}};
+  nlohmann::json result_9 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "string"}, {"maximum_size", 4}}}, {"name", "member_name"}};
+  nlohmann::json result_10 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "wstring"}, {"maximum_size", 5}}}, {"name", "member_name"}};
+  nlohmann::json result_11 = {{"comments", nlohmann::json::array()}, {"type", {{"name", "array"}, {"value_type", {{"name", "uint8"}}}, {"size", 123}}}, {"name", "member_name"}};
+
+  EXPECT_EQ(parse_member(test_string_1), result_1);
+  EXPECT_EQ(parse_member(test_string_2), result_2);
+  EXPECT_EQ(parse_member(test_string_3), result_3);
+  EXPECT_EQ(parse_member(test_string_4), result_4);
+  EXPECT_EQ(parse_member(test_string_5), result_5);
+  EXPECT_EQ(parse_member(test_string_6), result_6);
+  EXPECT_EQ(parse_member(test_string_7), result_7);
+  EXPECT_EQ(parse_member(test_string_8), result_8);
+  EXPECT_EQ(parse_member(test_string_9), result_9);
+  EXPECT_EQ(parse_member(test_string_10), result_10);
+  EXPECT_EQ(parse_member(test_string_11), result_11);
+}

